geomesa-core/src/main/scala/org/locationtech/geomesa/core/index/QueryPlanner.scala

```diff
@@ -154,28 +156,54 @@ case class QueryPlanner(schema: String,
     // Decode according to the SFT return type.
     // if this is a density query, expand the map
     if (query.getHints.containsKey(DENSITY_KEY)) {
-      accumuloIterator.flatMap { kv =>
-        DensityIterator.expandFeature(decoder.decode(kv.getValue.get))
-      }
+      adaptIteratorForDensityQuery(accumuloIterator, decoder)
     } else if (query.getHints.containsKey(TEMPORAL_DENSITY_KEY)) {
-      val timeSeriesStrings = accumuloIterator.map { kv =>
-        decoder.decode(kv.getValue.get).getAttribute(ENCODED_TIME_SERIES).toString
+      adaptIteratorForTemporalDensityQuery(accumuloIterator, returnSFT, decoder)
+    } else {
+      adaptIteratorForStandardQuery(accumuloIterator, query, decoder)
+    }
+  }
+
+
+  def adaptIteratorForStandardQuery(accumuloIterator: CloseableIterator[Entry[Key, Value]],
+                                    query: Query,
+                                    decoder: SimpleFeatureDecoder): CloseableIterator[SimpleFeature] = {
+    val features = accumuloIterator.map { kv =>
+      val ret = decoder.decode(kv.getValue.get)
+      val visibility = kv.getKey.getColumnVisibility
+      if(visibility != null && !EMPTY_VIZ.equals(visibility)) {
+        ret.getUserData.put(SecurityUtils.FEATURE_VISIBILITY, visibility.toString)
```

Comment:
Nice; I like the sequence of 'adapt' functions.  And that the vis is coming back to the user.